dotnet/src/SemanticKernel/Planning/SKContextExtensions.cs

```diff
@@ -1,48 +1,195 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.SemanticKernel.Memory;
 using Microsoft.SemanticKernel.Orchestration;
 using Microsoft.SemanticKernel.Orchestration.Extensions;
 using Microsoft.SemanticKernel.SkillDefinition;
+using static Microsoft.SemanticKernel.CoreSkills.PlannerSkill;
 
 namespace Microsoft.SemanticKernel.Planning;
 
 internal static class SKContextExtensions
 {
-    internal static string GetFunctionsManual(
+    internal const string MemoryCollectionName = "Planning.SKFunctionsManual";
+
+    internal const string ContextRegistrationKey = "Planning.SKFunctionsAreRemembered";
+
+    /// <summary>
+    /// Returns a string containing the manual for all available functions.
+    /// </summary>
+    /// <param name="context">The SKContext to get the functions manual for.</param>
+    /// <param name="semanticQuery">The semantic query for finding relevant registered functions</param>
+    /// <param name="config">The planner skill config.</param>
+    /// <returns>A string containing the manual for all available functions.</returns>
+    internal static async Task<string> GetFunctionsManualAsync(
         this SKContext context,
-        List<string>? excludedSkills = null,
-        List<string>? excludedFunctions = null)
+        string? semanticQuery = null,
+        PlannerSkillConfig? config = null)
     {
-        var functions = context.GetAvailableFunctions(excludedSkills, excludedFunctions);
-
-        return string.Join("\n\n",
-            functions.Select(
-                x =>
-                {
-                    var inputs = string.Join("\n", x.Parameters.Select(p => $"    -${p.Name}: {p.Description}"));
-                    return $"  {x.SkillName}.{x.Name}:\n    description: {x.Description}\n    inputs:\n{inputs}";
-                }));
+        config ??= new PlannerSkillConfig();
+        var functions = await context.GetAvailableFunctionsAsync(config, semanticQuery);
+
+        return string.Join("\n\n", functions.Select(x => x.ToManualString()));
```

Comment:
Would JsonSerializer.Serialize() work here?

Reply:
It might. Right now though this is what the prompt examples expect. The plan is to try a bunch of different strategies for R2.